python/app/input_widget.py

```diff
@@ -227,20 +228,364 @@ def keyPressEvent(self, event):
+            n_spaces += 4
+
+        # Add a new line plus the number of spaces in the previous line, plus any extra we might have added if a
+        # colon was found at the end of the line before.
+        new_line = "\n" + (" " * n_spaces)
+
+        # Remove the currently selected text, as usually editors delete the selection and move onto the new line.
+        self.textCursor().removeSelectedText()
+        cur.insertText(new_line)
+        # End our undo block.
+        cur.endEditBlock()
+
+    def block_comment_selection(self):
+        """
+        Either adds or removes comments from the selected line. If one line in the selection doesn't contain a comment
+         then it will add comments to all lines, otherwise it will remove comments from all lines.
+        :return: None
+        """
+
+        # Before attempting to alter the line, we should loop over the selected lines
+        # and check if any don't have a # at the start. If we find a line that doesn't then we
+        # will want to add commenting to all lines. If we don't find one then we want to remove the comments.
+        # Also during this loop we can find the lowest indentation point across all selected lines,
+        # and then use that as the index to insert the comment (if we are inserting)
+
+        cur = self.textCursor()
+
+        cur_pos, anchor, start, end = self._get_cursor_positions(cur)
+
+        # Loop over the lines from the bottom up
+        cur.setPosition(start)
+
+        add_comment = False
+        lowest_indent_index = None
+
+        hash_pattern = re.compile(r"^((?:[ \t]+)?)#")
+
+        while True:
+            line = cur.block().text()
+
+            # check to see if the line has a hash before the first character
+            if not hash_pattern.match(line):
+                add_comment = True
+
+            # find the lowest indent
+            indentation, rest_of_line = self._split_indentation(line)
+            before_first_char_index = len(indentation)
+            if (
+                lowest_indent_index is None
+                or before_first_char_index < lowest_indent_index
+            ):
+                lowest_indent_index = before_first_char_index
+
+            line_pos = cur.position()
+            # Now move up a line ready for the next loop and check we haven't gone beyond the start of the selection.
+            cur.movePosition(QtGui.QTextCursor.Down)
+
+            if line_pos == cur.position():
+                # moving down a line hasn't altered the position so we must be on the last time.
+                break
+
+            # Move to the start of the line before checking, so that we can be sure the cursor cannot be on the
+            # same line but after the end point.
+            cur.movePosition(QtGui.QTextCursor.StartOfLine)
+            if cur.position() > end:
+                # break out of the loop
+                break
+
+        def add_comment_to_line(line):
```

Comment:
I find that these methods defined in the middle of the code makes the code harder to read, especially since they have doc strings. Maybe make them `_`methods on the class instead?

Reply:
I've moved them to the top of the method if that is OK?
The trouble is that they are relying on variables defined in the outer scope, which won't be passed  through by the `_operate_on_selected_lines` method.

Reply:
I've also removed the doc strings, and added comments instead.